src/core/createModule.ts

```diff
@@ -0,0 +1,58 @@
+import type { ClientEvents, ClientOptions } from 'discord.js';
+import type { ZodTypeAny } from 'zod';
+import { z } from 'zod';
+
+import type { BotCommand, EventHandler } from '../types/bot';
+
+type InferredZodShape<Shape extends Record<string, ZodTypeAny>> = {
+  [K in keyof Shape]: Shape[K]['_type'];
+};
+
+interface Context<Env extends Record<string, ZodTypeAny>> {
+  env: InferredZodShape<Env>;
+}
+
+type ModuleFunction<Env extends Record<string, ZodTypeAny>, ReturnType> = (
+  context: Context<Env>,
+) => ReturnType;
+
+type EventHandlers = {
+  [K in keyof ClientEvents]?: EventHandler<K>;
+};
+
+type BotModule<Env extends Record<string, ZodTypeAny>> = {
+  env?: Env;
```

Comment:
Useless right now, we only store the env in the context, I think we should handle that when we really need it, in order to avoid adding more complexity